### PR TITLE
Add SQSListener argument resolver to extract the subject of an SNS me…

### PR DESCRIPTION
**Pull request by n-n-nebbl**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description

Support extracting of SNS subjects in SQS listeners using an annotation on a String parameter

I looked at the following PR and made the change accordingly to the current state of the code added there:
https://github.com/awspring/spring-cloud-aws/pull/898

## :bulb: Motivation and Context

This functionality was there in the old spring-cloud-aws library (2.x) and is making transitioning to the new library easier.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] I updated reference documentation to reflect the change
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

**Comment by n-n-nebbl** [comment]
It's the first time for me contributing to a bigger open-source project, so please tell me if I made some mistakes (also in the process of contributing besides adding the code)

If you agree that this functionality should be added, I will adding some tests and also have a look if I can make the code together with the SnsMessage PR having less duplicates.

@tomazfernandes Please have a look when you have time, so I know if I should put more effort into this.

**Comment by n-n-nebbl** [comment]
just a bump to maybe get an answer. maybe from one of you @maciejwalkowiak or @MatejNedic

File: spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/SnsJsonNode.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+			throw new MessageConversionException("Payload: '" + jsonString + "' does not contain a message", null);
+		}
+	}
+
+	public String getMessageAsString() {
+		return jsonNode.get("Message").asText();
+	}
+
+	public String getSubjectAsString() {
+		if (!jsonNode.has("Subject")) {
```
Comment (n-n-nebbl) [inline spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/SnsJsonNode.java:62]: Do SnsNotifications always have a `Subject`?
Then this could be moved to the `validate()` method.

Comment (n-n-nebbl) [reply spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/SnsJsonNode.java:62]: When sending a SNS message through AWS Console it is optional (https://docs.aws.amazon.com/sns/latest/dg/sns-publishing.html#publish-to-topic-aws-sdks)
In the `PublishRequest` class of the aws sdk sns package it is also described as optional in the javadoc.

So I think it is on the correct place already.

File: spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/SnsSubjectConverter.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * @author Alexander Nebel
+ * @since 3.3.0
+ */
+public class SnsSubjectConverter implements MessageConverter {
+
+	private final ObjectMapper jsonMapper;
+
+	public SnsSubjectConverter(ObjectMapper jsonMapper) {
```
Comment (tomazfernandes) [inline spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/SnsSubjectConverter.java:35]: Nit: We usually just name it `objectMapper`. The variable name is not an issue, but perhaps the error message would be clearer?

File: docs/src/main/asciidoc/sqs.adoc [hunk 3]
```diff
@@ -639,6 +639,16 @@ public void listen(@SnsNotificationMessage List<Pojo> pojos) {
 }
 ----
 
+Since 3.3.0 you can also retrieve the subject of the SNS message through the `@SnsNotificationSubject` annotation.
```
[the remark this answers is not in the document]
Comment (tomazfernandes) [reply docs/src/main/asciidoc/sqs.adoc:642]: Also, perhaps we should make it clear that lists are not supported?

BTW, any particular reason not to support Lists from the start?

File: spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/resolver/NotificationSubjectArgumentResolver.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		this.converter = new SnsSubjectConverter(jsonMapper);
+	}
+
+	@Override
+	public boolean supportsParameter(MethodParameter parameter) {
+		if (parameter.hasParameterAnnotation(SnsNotificationSubject.class)) {
+			if (ClassUtils.isAssignable(parameter.getParameterType(), String.class)) {
+				return true;
+			}
+			if (logger.isWarnEnabled()) {
```
Comment (tomazfernandes) [inline spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/resolver/NotificationSubjectArgumentResolver.java:51]: AFAIK modern logging implementations do not execute any logic e.g. placeholder resolution / String concatenation if the corresponding log level is not enabled, so I tend to favor cleaning up the code and not checking if the log level is enabled myself.

Just a nit though.

**Review by tomazfernandes — COMMENTED** [review]
Looks great @n-n-nebbl, thanks for the PR! 

Just left a few nits / suggestions for your consideration, but none of them are blockers.

File: docs/src/main/asciidoc/sqs.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (n-n-nebbl) [reply docs/src/main/asciidoc/sqs.adoc:642]: I just couldn't imagine the use case.
How would this look like?

```
void listen(@SnsNotificationSubject List<String> subjects, @SnsNotificationMessage List<String> messages) {
  ...
}
```
how to match the subject with the message then?
If this feature is needed I can implement it.

BTW: is the version correct? should it be 3.3.1?

File: spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/resolver/NotificationSubjectArgumentResolver.java (hunk 4 above)
Comment (n-n-nebbl) [reply spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/resolver/NotificationSubjectArgumentResolver.java:51]: static code analysis was complaining because I am calling `getMethodName(parameter)` even if the logging level is not activated which is true, but could also be ignored

File: docs/src/main/asciidoc/sqs.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomazfernandes) [reply docs/src/main/asciidoc/sqs.adoc]: Well, in theory it should be `3.4` since it's not a bugfix, but I think `3.3.1` is the safe bet. Unless @maciejwalkowiak has a different idea?

File: spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/resolver/NotificationSubjectArgumentResolver.java (hunk 4 above)
Comment (tomazfernandes) [reply spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/resolver/NotificationSubjectArgumentResolver.java:51]: Yeah, I guess that's a fair point.

Ideally `slf4j` could provide a method signature that received a lambda which would only be executed if logging was enabled at the proper level, but I don't think it's an option at this point.

We can leave it as is.

File: docs/src/main/asciidoc/sqs.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomazfernandes) [reply docs/src/main/asciidoc/sqs.adoc]: > I just couldn't imagine the use case.
> How would this look like?

Yup, fair point. We'd probably need a `Map<UUID, String>` instead of the list so it's possible to correlate, but I'm happy to leave this out of this PR and see if anyone requests that before implementing it.

[the remark this answers is not in the document]
Comment (maciejwalkowiak) [reply docs/src/main/asciidoc/sqs.adoc]: Unless its a breaking change it is safe to put 3.3.1

**Comment by n-n-nebbl** [comment]
From my point of view every comment is resolved and the PR is ready to get merged.

**Review by tomazfernandes — APPROVED** [review]
Thanks @n-n-nebbl, looking forward to more!

There are quite a few interesting SQS open issues if you're interested, please let me know if you'd like to pick any one up.

**Comment by tomazfernandes** [comment]
Just one late thought @n-n-nebbl - it might be interesting for us to have a `SnsMessage<MyType>` POJO that would contain both the deserialized message and the associated metadata from SNS.

So users could have either `@SnsNotificationMessage SnsMessage<MyPojo>` or `@SnsNotificationMessage List<SnsMessage<MyPojo>>`.

This way we could support getting a list of messages with their subjects and other metadata. We could also use it with the `ObjectMapper` internally instead of operating directly on JSONs.

Please let me know your thoughts and if you'd be interested in providing a PR for that. Thanks!

**Comment by n-n-nebbl** [comment]
> Just one late thought @n-n-nebbl - it might be interesting for us to have a `SnsMessage<MyType>` POJO that would contain both the deserialized message and the associated metadata from SNS.
> Please let me know your thoughts and if you'd be interested in providing a PR for that. Thanks!

That would be a nice solution!
I can provide a PR for it, but I do not know yet when I will have time for it so it could take some weeks.

**Comment by n-n-nebbl** [comment]
> Thanks @n-n-nebbl, looking forward to more!
> 
> There are quite a few interesting SQS open issues if you're interested, please let me know if you'd like to pick any one up.

I will have a look at them the next time I can spare some time coding :)
